qiskit/transpiler/passes/scheduling/time_unit_conversion.py

```diff
@@ -135,3 +175,74 @@ def _unified(unit_set: Set[str]) -> str:
             return "SI"
 
         return "mixed"
+
+
+_DURATION_KIND_NAME = {
+    Duration.dt: "dt",
+    Duration.ns: "ns",
+    Duration.us: "us",
+    Duration.ms: "ms",
+    Duration.s: "s",
+}
+
+
+class _EvalDurationImpl(expr.ExprVisitor[float]):
+    """Evaluates the expression to a single float result.
+
+    If ``dt`` is provided or all durations are already in ``dt``, the result is in ``dt``.
+    Otherwise, the result will be in seconds, and all durations MUST be in wall-time (SI).
+    """
+
+    __slots__ = ("dt", "has_dt", "has_si")
+
+    def __init__(self, dt=None):
+        self.dt = dt if dt is not None else 1
+        self.has_dt = False
+        self.has_si = False
+
+    def in_cycles(self):
+        """Returns ``True`` if units are 'dt' after visit."""
+        return self.has_dt or self.dt != 1
+
+    def visit_value(self, node, /) -> float:
+        if isinstance(node.value, float):
+            return node.value
+        if isinstance(node.value, Duration.dt):
+            if self.has_si and self.dt == 1:
+                raise TranspilerError(
+                    "Fail to unify time units in delays. SI units "
+                    "and dt unit must not be mixed when dt is not supplied."
+                )
```

Comment:
Technically I think the `self.dt == 1` check is unnecessary, since we both have SI and dt - we're going to be failing either way, right, even if dt isn't set?  (Not important to change.)

Reply:
If we have both SI and cycles, we're fine since we can convert, unless we don't have a `self.dt`.

This might just be hard to read. I changed `self.dt` to be `None` in this case, so now this reads `if self.has_si and self.dt is None:`

Reply:
oh of course yeah - that's the whole purpose of the pass lol